keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,946 @@
+If failure logs are concerned in this case (considering the logging container is sidecar) then the sidecar container will only be terminated when all app containers are in the failed state, so it must have captured the failure logs.
+
+The following are existing behaviors just mentioned here for clarity:
+
+- if all application containers exited successfully and the restart policy of the pod is "always" then restart the containers.
+
+- if all app containers are failed and the restart policy is "onFailure", then kubelet will keep trying to restart containers, and no point in terminating the pod
+
+We propose the following API field (spec.containers[].lifecycle.type) to identify a keystone container:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+    lifecycle:
+      type: Keystone
+  - name: sidecar
+    image: sidecar-image
+    command: ["do something to help my app"]
+```
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A Job consisting of two containers needs to be scheduled. It consists in a batch application that exits once the processing is done, and a sidecar sending logs to an external log server.
+
+As of today it's impossible to successfully terminate the Pod hosting this Job after the business processing has been done, because the log sending container will keep running unless some RPC is implemented between the two containers.
+
+By defining the batch processing container as a keystone container within the Pod, kubelet will properly handle this lifecycle.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the user identifies several (or all) containers as keystones, current logic doesn't determine the correct behavior which is whether to terminate the pod or keep it running. We mitigate this risk by allowing only 1 keystone container per pod for the moment to safely gather user feedback.
```

Comment:
I'm not sure I understand this risk. I think KEP is saying to wait for all keystone containers to complete and then terminate the rest? What part is not defined?